### PR TITLE
Improve the GenerateType code-actions menu experience.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
1. Put generate-type actions under a higher level "Generate Type" menu item if there are a lot of items in the code actions list.
2. If there aren't a lot of actions, then inline the items.
3. Simplify the wording of the "Generate type" actions to make the menu items cleaner.
4. Provide simple APIs to allow users to create top level menu actions with sub actions.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Reviewers: @balajikris, @basoundr, @brettfo, @davkean, @DustinCampbell, @dpoeschl, @Pilchie @jasonmalinowski

**Comment by balajikris** [comment]
Be sure to run the integration test suite before merging this to open. I have a feeling this may require updates to test actions and tests.

File: src/Features/Core/Portable/FeaturesResources.resx [hunk 1]
```diff
@@ -234,11 +234,11 @@
   <data name="GenerateLocal" xml:space="preserve">
     <value>Generate local '{0}'</value>
   </data>
-  <data name="GenerateForInNewFile" xml:space="preserve">
-    <value>Generate {0} for '{1}' in '{2}' (in new file)</value>
+  <data name="Generate_0_1_in_new_file" xml:space="preserve">
```
Comment (brettfo) [inline src/Features/Core/Portable/FeaturesResources.resx]: These underscores and numbers make my eyes bleed.  Is the IntelliSense not enough to visualize the resultant string?  E.g., the generated doc-comment that's of the form "Returns a string similar to Generate {0} for ..."

File: src/Workspaces/Core/Portable/CodeActions/CodeAction.cs [hunk 2]
```diff
@@ -284,57 +293,78 @@ public static CodeAction Create(string title, Func<CancellationToken, Task<Solut
...
-                _createChangedDocument = createChangedDocument;
+                _nestedActions = nestedActions;
                 _equivalenceKey = equivalenceKey;
             }
 
-            public override string Title
+            public sealed override string Title => _title;
+            public sealed override string EquivalenceKey => _equivalenceKey;
+
+            internal override bool IsInvokable => false;
```
Comment (balajikris) [inline src/Workspaces/Core/Portable/CodeActions/CodeAction.cs]: what is this Property for? Its `true` in the abstract base class, `false` here and `true` in all the derived classes. I guess I couldn't understand. Could you please explain. Alternatively doc comments would be nice too.

**Comment by CyrusNajmabadi** [comment]
@balajikris It's actually false in the code action we make for the "Action with nested children".  Effectively, we don't want to inline nested children if their parent is invokable itself.  We only can inline if the parent effectively serves only as a title and has no other purpose.

**Comment by CyrusNajmabadi** [comment]
I could just rename that to "CanInline" if you'd like that better :)

File: src/Features/Core/Portable/FeaturesResources.resx (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/FeaturesResources.resx]: I far prefer this form.  Now we dont' have this bizarre inconsistency between the identifier and the text it is replaced with.  There's a standard pattern for how you get the identifier given the format string.

Over in TS we just generated these identifiers automatically.  Thus, all you ever had to do was care about the format string.  Seems much nicer.   And much easier to grok in code.  "Oh... this needs X many arguments.  And here's where each goes.  And here's what the string will actally be"  All without needing to hover or anything.

File: src/Test/Diagnostics/project.lock.json [hunk 3]
```diff
@@ -113,255 +114,254 @@
         "lib/portable-net45+win8+wp8+wpa81/System.Composition.Runtime.XML",
         "lib/portable-net45+win8+wp8+wpa81/System.Composition.TypedParts.dll",
         "lib/portable-net45+win8+wp8+wpa81/System.Composition.TypedParts.XML",
-        "License-Stable.rtf",
-        "Microsoft.Composition.nuspec",
-        "package/services/metadata/core-properties/12e2c94035674b25b911285cc26e3260.psmdcp"
+        "package/services/metadata/core-properties/12e2c94035674b25b911285cc26e3260.psmdcp",
+        "[Content_Types].xml"
       ]
```
Comment (jmarolf) [inline src/Test/Diagnostics/project.lock.json]: Are these project.lock.json changes intentional?

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs [hunk 4]
```diff
@@ -160,11 +160,69 @@ public IEnumerable<SuggestedActionSet> GetSuggestedActions(ISuggestedActionCateg
...
+                // (but we only do this if the parent of the nested actions isn't invokable itself).
+                if (allActionSets.Sum(a => a.Actions.Count()) > 3)
+                {
+                    return allActionSets;
+                }
+
+                return allActionSets.Select(InlineActions).ToList();
+            }
+
+            private bool IsInlineable(ISuggestedAction action)
```
Comment (jmarolf) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs]: can be made static

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs [hunk 5]
```diff
@@ -160,11 +160,69 @@ public IEnumerable<SuggestedActionSet> GetSuggestedActions(ISuggestedActionCateg
...
+            {
+                var suggestedAction = action as SuggestedAction;
+                return suggestedAction != null &&
+                         !suggestedAction.CodeAction.IsInvokable &&
+                         suggestedAction.CodeAction.HasCodeActions;
+            }
+
+            private SuggestedActionSet InlineActions(SuggestedActionSet actionSet)
```
Comment (jmarolf) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs]: can be made static

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs [hunk 6]
```diff
@@ -160,11 +160,69 @@ public IEnumerable<SuggestedActionSet> GetSuggestedActions(ISuggestedActionCateg
...
+            {
+                if (!actionSet.Actions.Any(IsInlineable))
```
Comment (jmarolf) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSourceProvider.cs]: expanding to a lambda will cache the delegate conversion

**Comment by jmarolf** [comment]
:+1: looks good, but I don't understand the scope of the project.lock.json changes

File: src/Test/Diagnostics/project.lock.json (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Test/Diagnostics/project.lock.json]: Yes/no.  I mean, I didn't intentionally make them.  But I assume that any time these get updated, they need to get checked in.

Comment (jmarolf) [inline src/Test/Diagnostics/project.lock.json]: What scares me is that you just changed the contents of the files.  That shouldn't have changed project dependencies.

Comment (CyrusNajmabadi) [inline src/Test/Diagnostics/project.lock.json]: It looks like some boxes are sorting these values, and other boxes are not.  Pulling in @agocke to see if he knows why.

Comment (Pilchie) [inline src/Test/Diagnostics/project.lock.json]: I think @dpoeschl indicated that he that the 3.2RC version of the NuGet VS Extension applied sorting where the 3.1 version didn't.

File: src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs [hunk 7]
```diff
@@ -20,6 +22,11 @@ public partial class GenerateTypeTests : AbstractCSharpDiagnosticProviderBasedUs
                 null, new GenerateTypeCodeFixProvider());
         }
 
+        protected override IList<CodeAction> MassageActions(IList<CodeAction> codeActions)
```
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs]: Sounds so... dirty.

File: src/Test/Diagnostics/project.lock.json (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Test/Diagnostics/project.lock.json]: Is that something you need to explicitly get? Or do our standard command line tools pull them in?

Comment (Pilchie) [inline src/Test/Diagnostics/project.lock.json]: At the moment you'd need to go get the vsix separately and install it.

File: src/Features/Core/Portable/FeaturesResources.resx (hunk 1 above)
Comment (DustinCampbell) [inline src/Features/Core/Portable/FeaturesResources.resx]: I kind of like them. Given that they'll only ever be prefixes with a ```*Resources``` type, I think it'll look fine.

File: src/Features/Core/Portable/GenerateType/AbstractGenerateTypeService.cs [hunk 8]
```diff
@@ -72,7 +73,18 @@ protected AbstractGenerateTypeService()
                 var state = State.Generate((TService)this, semanticDocument, node, cancellationToken);
                 if (state != null)
                 {
-                    return GetActions(semanticDocument, node, state, cancellationToken);
+                    var actions = GetActions(semanticDocument, node, state, cancellationToken).ToList();
```
Comment (DustinCampbell) [inline src/Features/Core/Portable/GenerateType/AbstractGenerateTypeService.cs]: ```ToList()``` always makes me cry a little. It seems kind of necessary for the code below though.

File: src/Workspaces/Core/Portable/CodeActions/CodeAction.cs [hunk 9]
```diff
@@ -284,57 +293,78 @@ public static CodeAction Create(string title, Func<CancellationToken, Task<Solut
             return new SolutionChangeAction(title, createChangedSolution, equivalenceKey);
         }
 
-        internal class DocumentChangeAction : CodeAction
+        /// <summary>
+        /// Creates a top level code action with multiple code action choices a user can invoke.
+        /// </summary>
+        public static CodeAction Create(string title, IEnumerable<CodeAction> nestedActions, string equivalenceKey = null)
```
Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/CodeActions/CodeAction.cs]: Nice job getting the public API changes down to a single overload.

File: src/Workspaces/Core/Portable/CodeActions/CodeAction.cs (hunk 2 above)
Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/CodeActions/CodeAction.cs]: I suspect it refers to whether this code action can be invoked or not. For example, the parent code action of several nested code actions might not be invokable from the menu.

**Comment by DustinCampbell** [comment]
Minus some nits and the question around project.lock.json, I think this change looks good. Looking forward to trying it! :+1:
